### PR TITLE
Avatar

### PR DESCRIPTION
**Pull request by gnabyl**
Add avatar feature

### CODE REVIEW

File: lqdoj_backend/settings.py [hunk 1]
```diff
@@ -41,6 +41,7 @@
     'rest_framework',
     'rest_framework.authtoken',
     'corsheaders',
+    'django_cleanup'
```
Comment (tuankiet65) [inline lqdoj_backend/settings.py:44]: New dependency should be added to requirements.txt

File: lqdoj_backend/settings.py [hunk 2]
```diff
@@ -55,19 +56,14 @@
...
 ]
 
+REST_FRAMEWORK = {
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+        'rest_framework.authentication.TokenAuthentication',
+    )
+}
+
 CORS_ORIGIN_ALLOW_ALL = True
 CORS_ALLOW_CREDENTIALS = True
```
Comment (tuankiet65) [inline lqdoj_backend/settings.py]: `CORS_ALLOW_CREDENTIALS = True` (which implies`Access-Control-Allow-Credentials: true`) doesn't work when `CORS_ORIGIN_ALLOW_ALL = True` (implies `Access-Control-Allow-Origin = *`) since when `Access-Control-Allow-Origin: *` then `Access-Control-Allow-Credentials` must not be `true`. This should be addressed in a different commit though.

File: users/models.py [hunk 3]
```diff
@@ -1,3 +1,43 @@
+import os
+import time
+
+from PIL import Image
```
Comment (tuankiet65) [inline users/models.py:4]: New dependency should be added to requirements.txt (This dependency should be "Pillow" not "PIL")

File: lqdoj_backend/urls.py [hunk 4]
```diff
@@ -15,10 +15,15 @@
...
 
 urlpatterns = [
     path('admin/', admin.site.urls),
     path('users/', include('users.urls')),
     path('announcements/', include('announcements.urls')),
     path('tokens/', include('tokens.urls')),
 ]
+
+if settings.DEBUG:
+    urlpatterns += static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
```
Comment (tuankiet65) [inline lqdoj_backend/urls.py]: Not sure what you mean here, the media endpoint should be added at all time, regardless of whether the environment is debug or production.

File: users/forms.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+class UserUpdateForm(UserChangeForm):
+    class Meta:
+        model = User
+        fields = ("id", "username", "first_name", "last_name", "email")
+
+
+class ProfileUpdateForm(forms.ModelForm):
+    class Meta:
+        model = Profile
+        fields = ("avatar",)
```
Comment (tuankiet65) [inline users/forms.py]: Leave one newline at the end of the file

File: users/models.py [hunk 6]
```diff
@@ -1,3 +1,43 @@
...
+    time_obj = time.localtime(seconds_since_epoch)
+    return '%d-%d-%d_%d-%d-%d' % (
+        time_obj.tm_mday, time_obj.tm_mon, time_obj.tm_year, time_obj.tm_hour, time_obj.tm_min, time_obj.tm_sec)
+
+
+def get_file_path(instance, filename):
+    upload_to = os.path.join(AVATAR_FOLDER)
+    extension = filename.split('.')[-1]
+    filename = get_time_stamp() + "_" + instance.user.username
+    final_filename = '{}.{}'.format(filename, extension)
```
Comment (tuankiet65) [inline users/models.py]: There *could* be a chance of race condition if two avatar change requests arrive at the same second. Might as well use uuid.uuid4() since the generated UUID is guaranteed to be unique.

File: users/models.py [hunk 7]
```diff
@@ -1,3 +1,43 @@
...
+    seconds_since_epoch = time.time()
+    # Convert seconds since epoch to struct_time
+    time_obj = time.localtime(seconds_since_epoch)
+    return '%d-%d-%d_%d-%d-%d' % (
+        time_obj.tm_mday, time_obj.tm_mon, time_obj.tm_year, time_obj.tm_hour, time_obj.tm_min, time_obj.tm_sec)
+
+
+def get_file_path(instance, filename):
+    upload_to = os.path.join(AVATAR_FOLDER)
+    extension = filename.split('.')[-1]
```
Comment (tuankiet65) [inline users/models.py]: Use `os.path.splitext(filename)[1]` intead of `filename.split`

File: users/models.py [hunk 8]
```diff
@@ -1,3 +1,43 @@
...
+    avatar = models.ImageField(default='default.jpg', upload_to=get_file_path)
+
+    def __str__(self):
+        return "User profile: " + self.user.username
+
+    def save(self, force_insert=False, force_update=False, using=None,
+             update_fields=None):
+        super().save()
+        img = Image.open(self.avatar.path)
+        if (img.height > 150) or (img.width > 150):
```
Comment (tuankiet65) [inline users/models.py]: 150x150 is pretty small though, how about 600x600?

File: users/serializers.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        model = Profile
+        fields = ('avatar',)
+
+
+class UserSerializer(serializers.ModelSerializer):
+    profile = ProfileSerializer()
+
+    class Meta:
+        model = User
+        fields = ("id", "username", "first_name", "last_name", "email", "date_joined", "profile")
```
Comment (tuankiet65) [inline users/serializers.py]: One newline at the end of the file.

File: users/models.py [hunk 10]
```diff
@@ -1,3 +1,43 @@
...
-# Create your models here.
\ No newline at end of file
+from lqdoj_backend.settings import AVATAR_FOLDER
+
+
+def get_time_stamp():
+    # Get the seconds since epoch
+    seconds_since_epoch = time.time()
+    # Convert seconds since epoch to struct_time
+    time_obj = time.localtime(seconds_since_epoch)
```
Comment (tuankiet65) [inline users/models.py]: If you're insisted on using the current date and time, then a better way to get the current time is `datetime.datetime.now()` and `datetime.datetime.utcnow()`

File: users/models.py [hunk 11]
```diff
@@ -1,3 +1,43 @@
...
+
+class Profile(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE, to_field="username")
+    avatar = models.ImageField(default='default.jpg', upload_to=get_file_path)
+
+    def __str__(self):
+        return "User profile: " + self.user.username
+
+    def save(self, force_insert=False, force_update=False, using=None,
+             update_fields=None):
```
Comment (tuankiet65) [inline users/models.py]: Unused arguments? Or they should be passed onto super().save()?

File: users/views.py [hunk 12]
```diff
@@ -65,6 +46,14 @@ def create(self, request, *args, **kwargs):
     """
 
     def partial_update(self, request, *args, **kwargs):
+        # Return 401 if no token found
+        if request.auth is None:
+            return Response(status=HTTP_401_UNAUTHORIZED)
+
+        # Return 401 if token doesn't match with the lookup value
+        if request.user.username != kwargs[self.lookup_field]:
+            return Response(status=HTTP_401_UNAUTHORIZED)
```
Comment (tuankiet65) [inline users/views.py]: 401 is for when a token isn't included in the request. Since a token mismatch means someone is trying to modify someone else's profile, this should return a 403 (Forbidden)

File: users/views.py [hunk 13]
```diff
@@ -77,11 +66,23 @@ def partial_update(self, request, *args, **kwargs):
...
+        # Return 401 if no token found
+        if request.auth is None:
             return Response(status=HTTP_401_UNAUTHORIZED)
-        form = CustomUserChangeForm(data=json.loads(request.body), instance=request.user)
-        if form.is_valid():
-            form.save()
+
+        # Return 401 if token doesn't match with the lookup value
+        if request.user.username != kwargs[self.lookup_field]:
+            return Response(status=HTTP_401_UNAUTHORIZED)
```
Comment (tuankiet65) [inline users/views.py]: Same as above

File: users/signals.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from users.models import Profile
+
+
+@receiver(post_save, sender=User)
+def create_profile(sender, instance, created, **kwargs):
+    if created:
+        Profile.objects.create(user=instance)
```
Comment (tuankiet65) [inline users/signals.py:11]: Might as well integrate this directly into the view logic, instead of using signals?

**Review by tuankiet65 — COMMENTED** [review]
Also
* please .gitignore the media folder (except for the default avatar file)
* probably choose a different, more appropriate default avatar?

File: users/signals.py (hunk 14 above)
Comment (gnabyl) [reply users/signals.py:11]: For me it's more clear to do something like this. In this file you can see and keep track all signals handler (for example if we have more models which are associated with User models), instead of looking in different function implementation in views.py file.

File: users/models.py (hunk 11 above)
Comment (gnabyl) [reply users/models.py]: In fact, it's overriding. It's obligate to write an override function with the same args with the save function from super
I can replace it with `def save(self, **kwargs):` but didn't test. Anyway we have to pass the same args when override

File: lqdoj_backend/urls.py (hunk 4 above)
Comment (gnabyl) [reply lqdoj_backend/urls.py]: Check the documentation here: 
https://docs.djangoproject.com/en/2.2/howto/static-files/#serving-static-files-during-development
Bref, it's recommended to implement another way of serving static files in production

File: users/models.py (hunk 3 above)
Comment (gnabyl) [reply users/models.py:4]: I have added all dependencies to dependencies.txt in master branch
I tent to add it here but by some mistakes during the git work, but nevermind
Anyway, by another mistake, I have committed the pipfile too, must I delete it?

File: lqdoj_backend/settings.py (hunk 1 above)
Comment (tuankiet65) [reply lqdoj_backend/settings.py:44]: Resolved

File: users/models.py (hunk 3 above)
Comment (tuankiet65) [reply users/models.py:4]: Resolved.

> Anyway, by another mistake, I have committed the pipfile too, must I delete it?

No, it's necessary in fact.

File: lqdoj_backend/urls.py (hunk 4 above)
Comment (tuankiet65) [reply lqdoj_backend/urls.py]: All right, that's what I thought at first. Maybe you should add a TODO there like

```python
# TODO: Implement another storage system suitable for production use
```

File: users/models.py (hunk 11 above)
Comment (tuankiet65) [reply users/models.py]: Probably something like this would work (reference: https://stackoverflow.com/a/4269828/1851156)
```python
def save(self, *args, **kwargs):
    super().save(*args, **kwargs)
```

File: users/signals.py (hunk 14 above)
Comment (tuankiet65) [reply users/signals.py:11]: All right.

File: .gitignore [hunk 15]
```diff
@@ -63,6 +62,9 @@ instance/
 # Scrapy stuff:
 .scrapy
 
+# Media stuff:
+media/avatars
```
Comment (tuankiet65) [inline .gitignore]: Might as well just filter out the entire `media` folder, then use `!` to include specific files. For example:

```gitignore
media/
!media/default.jpg
```

Also how about moving `default.jpg` to `media/avatars`? Seems more appropriate to me.

File: users/models.py (hunk 11 above)
Comment (gnabyl) [reply users/models.py]: Tested, it worked. 
In fact, this comment help me to realize a bug that will cause HTTP_500 when user try to update information. What a Quack debug!

File: users/models.py (hunk 8 above)
Comment (gnabyl) [reply users/models.py]: We need to store a big avatar for what?

Comment (tuankiet65) [reply users/models.py]: Showing the avatar in the profile page for example? At least we should store a high-resolution version, then in the future we could create smaller versions from the original for use in places where the avatar is small.

File: .gitignore [hunk 16]
```diff
@@ -63,6 +62,13 @@ instance/
 # Scrapy stuff:
 .scrapy
 
+# Media stuff:
+# Media stuff:
```
Comment (tuankiet65) [inline .gitignore]: Duplicate lines

File: lqdoj_backend/settings.py [hunk 17]
```diff
@@ -55,19 +56,14 @@
...
-    'accept-encoding',
-    'authorization',
-    'content-type',
-    'dnt',
-    'origin',
-    'user-agent',
-    'x-csrftoken',
-    'x-requested-with'
-]
+CORS_ALLOW_CREDENTIALS = False
```
Comment (tuankiet65) [inline lqdoj_backend/settings.py]: The default value is False so we don't have to explicitly set it to False (reference: https://github.com/adamchainz/django-cors-headers#cors_allow_credentials)

File: users/models.py [hunk 18]
```diff
@@ -1,3 +1,38 @@
...
+
+class Profile(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE, to_field="username")
+    avatar = models.ImageField(default='avatars/default.jpg', upload_to=get_file_path)
+
+    def __str__(self):
+        return "User profile: " + self.user.username
+
+    def save(self, *args, **kwargs):
+        super().save()
```
Comment (tuankiet65) [inline users/models.py]: Shouldn't we pass *args and **kwargs to super().save()? And also I think we should save the image first before saving the Profile.

File: users/views.py [hunk 19]
```diff
@@ -65,6 +47,14 @@ def create(self, request, *args, **kwargs):
     """
 
     def partial_update(self, request, *args, **kwargs):
+        # Return 401 if no token found
+        if request.auth is None:
+            return Response(status=HTTP_401_UNAUTHORIZED)
+
+        # Return 401 if token doesn't match with the lookup value
```
Comment (tuankiet65) [inline users/views.py]: Comment says 401 while the code returns 403

File: users/views.py [hunk 20]
```diff
@@ -77,11 +67,23 @@ def partial_update(self, request, *args, **kwargs):
...
     def update(self, request, *args, **kwargs):
-        if (request.auth is None) or (request.user.username != kwargs[self.lookup_field]):
+        # Return 401 if no token found
+        if request.auth is None:
             return Response(status=HTTP_401_UNAUTHORIZED)
-        form = CustomUserChangeForm(data=json.loads(request.body), instance=request.user)
-        if form.is_valid():
-            form.save()
+
+        # Return 401 if token doesn't match with the lookup value
```
Comment (tuankiet65) [inline users/views.py]: Same as above

File: users/models.py (hunk 18 above)
Comment (gnabyl) [reply users/models.py]: Passing or not, it's not matter, fixed to pass

File: users/models.py (hunk 8 above)
Comment (gnabyl) [reply users/models.py]: I think 150 is enough, anyway changed to 500x500, it's just decreasing the speed.
